ykrt/src/compile/jitc_yk/codegen/x86_64/lsregalloc.rs

```diff
@@ -0,0 +1,1099 @@
+//!
+//! ```rust,ignore
+//! match binop {
+//!   BinOp::Add => {
+//!     let size = lhs.byte_size(self.m);
+//!     let [lhs_reg, rhs_reg] = self.ra.get_gp_regs(
+//!       &mut self.asm,
+//!       iidx,
+//!       [RegConstraint::InputOutput(lhs), RegConstraint::Input(rhs)],
+//!     );
+//!     match size {
+//!       1 => dynasm!(self.asm; add Rb(lhs_reg.code()), Rb(rhs_reg.code())),
+//!       ...
+//!     }
+//! ```
+//!
+//! This says "return two x64 registers: `lhs_reg` will take a value as input and later contain an
+//! output value (clobbering the input value, which will be spilled if necessary); `rhs_reg` will
+//! take a value as input (and mustn't clobber it)". Those registers can then be used with dynasmrt
+//! as one expects.
+//!
+//! The allocator keeps track of which registers have which trace instruction's values in and of
+//! where it has spilled an instruction's value: it guarantees to spill an instruction to at most
+//! one place on the stack.
+
+use crate::compile::jitc_yk::{
+    codegen::abs_stack::AbstractStack,
+    jit_ir::{Const, ConstIdx, FloatTy, InstIdx, Module, Operand, Ty},
+};
+use dynasmrt::{
+    dynasm,
+    x64::{
+        Assembler, {Rq, Rx},
+    },
+    DynasmApi, Register,
+};
+use std::marker::PhantomData;
+
+/// The complete set of general purpose x64 registers, in the order that dynasmrt defines them.
+/// Note that large portions of the code rely on these registers mapping to the integers 0..15
+/// (both inc.) in order.
+pub(crate) static GP_REGS: [Rq; 16] = [
+    Rq::RAX,
+    Rq::RCX,
+    Rq::RDX,
+    Rq::RBX,
+    Rq::RSP,
+    Rq::RBP,
+    Rq::RSI,
+    Rq::RDI,
+    Rq::R8,
+    Rq::R9,
+    Rq::R10,
+    Rq::R11,
+    Rq::R12,
+    Rq::R13,
+    Rq::R14,
+    Rq::R15,
+];
+
+/// How many general purpose registers are there? Only needed because `GP_REGS.len()` doesn't const
+/// eval yet.
+const GP_REGS_LEN: usize = 16;
+
+/// FIXME: A temporary scratch register we guarantee not to use elsewhere. Every use of this is a
+/// hack.
+static WR0: Rq = Rq::R12;
+
+/// The complete set of floating point x64 registers, in the order that dynasmrt defines them.
+/// Note that large portions of the code rely on these registers mapping to the integers 0..15
+/// (both inc.) in order.
+static FP_REGS: [Rx; 16] = [
+    Rx::XMM0,
+    Rx::XMM1,
+    Rx::XMM2,
+    Rx::XMM3,
+    Rx::XMM4,
+    Rx::XMM5,
+    Rx::XMM6,
+    Rx::XMM7,
+    Rx::XMM8,
+    Rx::XMM9,
+    Rx::XMM10,
+    Rx::XMM11,
+    Rx::XMM12,
+    Rx::XMM13,
+    Rx::XMM14,
+    Rx::XMM15,
+];
+
+/// How many floating point registers are there? Only needed because `FP_REGS.len()` doesn't const
+/// eval yet.
+const FP_REGS_LEN: usize = 16;
+
+/// The set of general registers which we will never assign value to. RSP & RBP are reserved by
+/// SysV.
+///
+/// FIXME: R12 is a temporary hack because it's the "WR0" hack.
+static RESERVED_GP_REGS: [Rq; 3] = [Rq::RSP, Rq::RBP, Rq::R12];
```

Comment:
I wonder why you tell the allocator about the existence of the reserved registers at all. Is there a reason?

Reply:
You just don't include it in the set of registers that are available for allocation.